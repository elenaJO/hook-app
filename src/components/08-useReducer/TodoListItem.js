import React from 'react'
import PropTypes from 'prop-types';

export const TodoListItem = ({todo, index, handleToogle, handleDelete}) => {
	return (
		<li key={todo.id} className="list-group-item">
			<p
				onClick={ () => handleToogle(todo.id) }
				className={ `${ todo.done && 'complete' }` }
			>
				{ index + 1}. { todo.desc }
			</p>
			<button
				onClick={ () => handleDelete(todo.id) }
				className="btn btn-danger"
				type="button"
			>Borrar</button>
		</li>
	)
}

TodoListItem.protoTypes = {
	index: PropTypes.number.isRequired,
	todo: PropTypes.array.isRequired,
	handleToogle: PropTypes.func.isRequired,
	handleDelete: PropTypes.func.isRequired,
}