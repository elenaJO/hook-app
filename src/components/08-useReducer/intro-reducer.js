const initialState = [{
	id: 1,
	todo: 'Comprar pan',
	done: false,
}];

//siempre tiene q retornar un state
//no usar push ya q modifica el state
const todoReducer = (state = initialState, action) => {
	if (action?.type === 'agregar') {
		//no modifico el state inicial solo devuelvo agregando el nuevo
		// solo debe resolver con los argumentos q le mando el state y action
		return [...state, action.payload];
	}
	return state;
}

let todos = todoReducer();

const newTodo = {
	id: 2,
	todo: 'Comprar leche',
	done: false,
}

//es un estandar que tenga el campo type q tipo de accion es
// es un estandra que se llame payload
const agregarTodoAction = {
	type: 'agregar',
	payload: newTodo,
}

todos = todoReducer(todos, agregarTodoAction);

console.log(todos);