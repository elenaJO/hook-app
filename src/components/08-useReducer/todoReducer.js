export const todoReducer = (state = [], action) => {
	switch (action.type) {
		case 'add':
			return [...state, action.payload];
		//ya no va el break por q no necesito ver las otras condicionales	
		//siempre el default ya q se llama cuando se inicializa
		case 'delete':
			//el filter devuelve un nuevo array no muta
			return state.filter(todo => todo.id !== action.payload);
		case 'toogle':
		//con el ternario hacemos un return implicito ya q siempre retorna algo
			return state.map(todo => 
				(todo.id === action.payload)
					? { ...todo, done: !todo.done }
					: todo
			);
		default:
			return state;
	}
}