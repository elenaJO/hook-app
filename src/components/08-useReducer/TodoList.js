import React from 'react'
import './style.css';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleToogle, handleDelete }) => {
	return (
		<ul className="list-group list-group-flush">
			{
				todos.map((todo, index) => (
					// <li key={todo.id} className="list-group-item">
					// 	<p
					// 		onClick={ () => handleToogle(todo.id) }
					// 		className={ `${ todo.done && 'complete' }` }
					// 	>
					// 			{ index + 1}. { todo.desc }
					// 	</p>
					// 	<button
					// 		onClick={ () => handleDelete(todo.id) }
					// 		className="btn btn-danger"
					// 		type="button"
					// 	>Borrar</button>
						
					// </li>
					<TodoListItem
						key={todo.id}
						todo={todo}
						index={index}
						handleToogle={ handleToogle }
						handleDelete={ handleDelete }
					/>
				))
			}
		</ul>
	)
}
