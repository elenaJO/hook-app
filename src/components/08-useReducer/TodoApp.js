import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
// import { useForm } from '../../hooks/useForm';
import './style.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

// const initialState = [{
// 	id: new Date().getTime(),
// 	desc: 'Aprender React',
// 	done: false,
// }];

const init = () => {
	return JSON.parse(localStorage.getItem('todos')) ||Â [];
}

export const TodoApp = () => {

	//initialState -> es un estado inicial
	//init -> es una funcion para inicializar el state en caso sea como procesado o haga varias acciones, ayuda a react a computar estado incial para q no se ejecuta cada rato cuando aya cambios
	//dispatch -> disparar mis acciones hacia mi reducer, es para q react se entere para q vuelva a redibujar en pantalla
	const [ todos, dispatch ] = useReducer(todoReducer, [], init);
	// const [ { description }, handleInputChange, reset] = useForm({
	// 	description: '',
	// });

	useEffect(() => {
		localStorage.setItem('todos', JSON.stringify(todos));
	}, [todos]);

	// const handleSubmit = (e) => {
	// 	e.preventDefault();

	// 	if (description.trim().length <=1) {
	// 		return;
	// 	}
	// 	const newTodo = {
	// 		id: new Date().getTime(),
	// 		desc: description,
	// 		done: false, 
	// 	};

	// 	const action = {
	// 		type: 'add',
	// 		payload: newTodo,
	// 	}

	// 	//dispatch es para disparar la accion

	// 	//se puede enviar como argumentos a otros 
	// 	//componentes hijos y ese dispatch sabe q pertenece al useReducer, 
	// 	//por q los objetos son pasados como referencia en javascript
	// 	dispatch(action);
	// 	reset();
	// }

	const handleDelete = (id) => {
		const action = {
			type: 'delete',
			payload: id,
		};
		dispatch(action);
	}

	const handleToogle = (id) => {
		dispatch({
			type: 'toogle',
			payload: id,
		});
	}

	const handleAdd = (newTodo) => {
		dispatch({
			type: 'add',
			payload: newTodo,
		});
	}
	
	return (
		<div>
			<h1>TodoApp ({ todos.length })</h1>
			<hr />
			<div className="row">
				<div className="col-7">
					<TodoList 
						todos={ todos }
						handleToogle={ handleToogle }
						handleDelete= { handleDelete }
					/>
					{/* <ul className="list-group list-group-flush">
						{
							todos.map((todo, index) => (
								<li key={todo.id} className="list-group-item">
									<p
										onClick={ () => handleToogle(todo.id) }
										className={ `${ todo.done && 'complete' }` }
									>
											{ index + 1}. { todo.desc }
									</p>
									<button
										onClick={ () => handleDelete(todo.id) }
										className="btn btn-danger"
										type="button"
									>Borrar</button>
								</li>
							))
						}
					</ul> */}
				</div>
				<div className="col-5">
					{/* <h4>Agregar TODO</h4>
					<hr/>
					<form onSubmit={ handleSubmit }>
						<input
							type="text"
							name="description"
							className="form-control"
							placeholder="Aprender ..."
							autoComplete="off"
							value={ description }
							onChange={ handleInputChange }
						/>
						<button
							type="submit"
							className="btn btn-outline-primary mt-2 w-100">
							Agregar
						</button>
					</form> */}
					<TodoAdd handleAdd={handleAdd}/>
				</div>
			</div>
		</div>
	)
}
