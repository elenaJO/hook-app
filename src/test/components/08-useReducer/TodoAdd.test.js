import { shallow } from 'enzyme';
import React from 'react';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';

describe('Pruebas en el componente <TodoAdd/>', () => {
	const handleAdd = jest.fn();
	const wrapper = shallow(
		<TodoAdd
			handleAdd={ handleAdd }
		/>
	)

	test('debe mostrarse correctamente', () => {
		expect(wrapper).toMatchSnapshot();
	})

	test('NO debe llamar a la funcion handleAdd', () => {
		//otra forma de simular es, da una funcion y lo podemo llamar
		const formSubmit = wrapper.find('form').prop('onSubmit');
		formSubmit({ preventDefault(){} });
		//no se llamo ninguna vez
		expect(handleAdd).toHaveBeenCalledTimes(0);
	})

	test('debe llamar la funcion handleAdd', () => {
		const value = 'Aprender React';
		wrapper.find('input').simulate('change', 
			{ 
				target:  { name: 'description', value }
			});
		const formSubmit = wrapper.find('form').prop('onSubmit');
		formSubmit({ preventDefault(){} });
		expect(handleAdd).toHaveBeenCalledTimes(1);
		expect(handleAdd).toHaveBeenCalledWith(expect.any(Object)); //esta bien pero tmb puedo mandar objeto vacio no es muy preciso
		expect(handleAdd).toHaveBeenCalledWith({
			id: expect.any(Number), //espera cualquier numero
			desc: value,
			done: false,
		});
		expect(wrapper.find('input').prop('value')).toBe(''); //vemos los efectos del reset, que se limpie el input
	})
})