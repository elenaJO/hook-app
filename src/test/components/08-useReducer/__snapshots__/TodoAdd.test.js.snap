// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoAdd/> debe mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender ..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-2 w-100"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
